### DISCUSSION TITLE
Select Communication Speed (DECSCS)

### DISCUSSION THREAD

**Discussion by NuSkooler** [mkrueger/icy_term#19] [General]
I'd like to add support for "Select Communication Speed" (DECSCS): https://vt100.net/docs/vt510-rm/DECSCS.html in the manor of SyncTERM and some others (https://github.com/NuSkooler/ansi-bbs-utils/blob/master/docs/reference/cterm.txt)

Any hints as to where to look into for implementing this? `AnsiParser` portion is obvious, but to do this, there will need to be some 'pause' between characters received and displayed. Thoughts?

**Comment by mkrueger** [comment]
I wouldn't put that into the AnsiParser - the ANSI parser needs to parse that and store it maybe in the terminal information of the buffer. 

But the throttling can be done on the com layer. Adding the communication speed as an option on the UI level makes sense as well btw.

**Comment by mkrueger** [comment]
Note: Added the command support here:
https://github.com/mkrueger/icy_engine/commit/b16eedf34a46f72c1f2fbeca3079fc73ca28dd93

Now the terminal just needs to pickup that and the communication thread needs to throttle that somehow. I'm not sure if I should stick to tokio. It's really hard to get it working with wasm. At some point it would be cool to support web, ios and android as platform.

**Comment by NuSkooler** [comment]
@mkrueger Awesome! Yep, your implementation in AnsiParser looked like what I was thinking. I'll look into getting the throttling and a UI element in place.

**Reply by mkrueger** [reply]
Basically it's just a sleep  in the com layer. It needs the baud rate setting and measure how much bytes there are. Maybe using a decorator for that?

**Reply by NuSkooler** [reply]
I poked around a tad - in the com layer, reads are buffered so sleeping there doesn't make much sense, unless I'm misunderstanding what you're suggesting. Within `update_state()` character-by-character is drawn, but millisecond sleeping there is a no-go I believe.

**Reply by mkrueger** [reply]
Take a look at the no_tokio branch (IDK if I've merged it to main if so it doesn't exists).

Maybe the communication thread is the easiest place to handle it directly in the 
"open_connection"  method in main_window_mod.rs (I really need to work on naming)

There is the main loop. 

In handle.read_data() the bytes are read from the com layer. and with tx.send the data is send to the ui thread.
That needs to be limited there. A VecDeque where the data is stored and if it's not empty the head needs to be drained according to the baud rate.

Should be fairly easy. 

ps. I think it's good to remove tokio - it's a nice library for server side applications but for handling one connection it's maybe a bit overkill and I still try to play around with the possibility to run in a browser, ios & android. The only feature request I got so far is: "make the thing running on my tablet"  :)

**Reply by NuSkooler** [reply]
Grabbed that branch, definitely easier to deal with here. Hopefully I'll have some time tonight!

**Reply by mkrueger** [reply]
I don't think I'll implement wasm - I would've to switch the rendering engine from glow to wgpu - maybe  later when I know a bit about wgpu. But atm I don't have much enthusiasm switching the ui rendering layer. Was a pain in the ass getting glow working.

Same with SSH - no library I tested works. It always doesn't get something right with the crypt algorithms. Something I really don't want to know/handle on user level :(.

I'll tidy up the up&download protocols next that's a way better time investment. Now with the simpler communication layer unit tests are working again.

btw. do you see any differences between the tokio and non tokio layer?

**Reply by NuSkooler** [reply]
> I don't think I'll implement wasm - I would've to switch the rendering engine from glow to wgpu - maybe later when I know a bit about wgpu. But atm I don't have much enthusiasm switching the ui rendering layer. Was a pain in the ass getting glow working.

Damn, I think Wasm support would be amazing: In-browser with GPU, mobile, desktop client, yum! What portions need re-worked?

> Same with SSH - no library I tested works. It always doesn't get something right with the crypt algorithms. Something I really don't want to know/handle on user level :(.

I know this well, so I'd love to help. I haven't had time to learn your library fully here quite yet, but if you have a branch with SSH I can look at, I can look into handshakes, cipher suites, etc.

RE non-tokio, not yet. So far, so good!

**Comment by mkrueger** [comment]
> Damn, I think Wasm support would be amazing: In-browser with GPU, mobile, desktop client, yum! What portions need re-worked?

I render everything on the graphics card with shaders but the shaders need to be feeded somehow to the gfx card. I had a 50:50 chance - egui supports glow and wgpu. Seemed to me that glow is the default. 

But it doesn't compile for wasm atm- I need to do a deeper look on that. Maybe it's possible to use glow somewhow. But much work I did in the last ~2 weeks was targeted to get to the web & mobile devices. 

From the API pov wgpu looks even better - I should've used that. First thing I would do is to explore my possibilities - doing an extra project and see if it's possible to get it to run.

A simple 3 line shader that paints everything white or so would be enough.

The shader pipeline I did is very basic - create textures - I encode the buffer as a special texture that contains the char data & attributes - everything is calculated on the graphics card.

Font & Palette is a texture. As well as all sixels. So it's 'just' a matter of getting the textures created and the shader programs to run. But I've not much experience with that - it's the first time I do something with fragment shaders. May even have done many bad things :/.

But it's nice to have smooth scrolling or CRT effects - I made a new options panel where it's possible to play around with it a bit. But it's nowhere on the level of let's say "Cool Retro Term".  


There are tons of interesting and funny shader effects out there, only problem is that many of them are GPL and I think I stick with Apache 2 as license.

**Reply by mkrueger** [reply]
Maybe it makes sense to explore:

https://github.com/not-fl3/egui-miniquad

Looks promising.

**Comment by NuSkooler** [comment]
What's your thoughts on communicating the current baud rate (which can be set dynamically the parser) to the thread in `open_connection()`? If we do it outside of the thread, only the user setting (ie: phonebook entry) could work.

**Reply by mkrueger** [reply]
In open_connection it's surely one suitable point for a user setting. Communicating with the thread can be done with a message - like SendData, etc. adding a "ChangeBaudRate" should be easy.

I'll add a callback action for SetBaudRate as well.

**Comment by NuSkooler** [comment]
WIP here: https://github.com/mkrueger/icy_term/pull/23/files

With a callback, I'll be able to wire up the channel to send the info

**Reply by mkrueger** [reply]
Added a callback: https://github.com/mkrueger/icy_term/commit/28c432b7bcdaa3702da6ded07aeea5b7b2d3b588

**Reply by mkrueger** [reply]
However need to glue it together. And I really, really need to refactor things - had a code duplication. Rust is sometimes a bit too strict so I duplicated code there.

**Reply by mkrueger** [reply]
Ok I stop on the baud thing atm otherwise we'll collide too easily. I've to fix up/download protocols.

**Comment by mkrueger** [comment]
Com layer is still a mess  - the 'Connection' should be merged with the communication thread :/.

The connection is from pre tokio and the thread is from tokio. Having a thread makes sense. Blocking the UI thread on tcp is a bad idea. Maybe it even makes sense to have a thread for parsing - however parsing is very fast.

I should finish the up/download UI - atm it's an unusuable place holder but I'm too tempted to work on icy view instead - a small ansi viewing tool. I could port the new rendering layer and maybe work on getting that compiled for wasm instead :).

It's kind of a 10 times simpler project.

**Comment by mkrueger** [comment]
Seems we need more threads :).
One for sending and one for receiving. ATM the receiving part can block the sending if the baud rate is too slow. 

The receiving threads needs a way to be canceled/restarted or paused because the up/download protocols take over the line directly. 

Or the bps approach needs to be non blocking - may be easier.  My struct "struct Blink" may work, it's a very basic approach - I used that for caret & background blinking.

**Reply by NuSkooler** [reply]
I wondered about that - is this something you want to tackle, or should I poke at it?

**Reply by mkrueger** [reply]
I can take a look. I've played around with the web backend. Seems that some slight changes make it compile for wasm.

Got it running today… however doesn't show anything - which limits the user experience a bit :).

I need to figure out how to debug a wasm web application.  When I show the default eframe template app it works. Very strange - I suppose something is crashing. 
0 xp on wasm apps on my side so far. And less with web :)

**Reply by NuSkooler** [reply]
You might have some better luck running outside the browser for at least initial development using a runtime like Wasmtime, Wasmer, etc. At least under Linux, you can mostly just use standard debugging that way.

**Reply by mkrueger** [reply]
I got it running :). Rendering of the terminal is atm broken but many obstacles are fixed.

Glow runs on the web - but I needed to do some workarounds. May've used the wrong OpenGL calls here and there. There are slight differences between web and desktop - unfortunately.

**Reply by NuSkooler** [reply]
I just pulled the latest from `no_tokio` - stranger rendering issue: Characters missing - see SS:
![image](https://github.com/mkrueger/icy_term/assets/1509339/a578ed22-8394-4499-9b87-0e20736a5284)

**Reply by NuSkooler** [reply]
Another - these look to be missing '@' symbol (e.g. in front of 'to' name): 
![image](https://github.com/mkrueger/icy_term/assets/1509339/0b1e63d6-2157-44aa-bceb-2cd00db5f545)

**Reply by NuSkooler** [reply]
One other interesting tidbit: Debug logs showing in the terminal itself:
![image](https://github.com/mkrueger/icy_term/assets/1509339/d41aef02-9550-4dfa-a2b4-8cd9ae127f6a)

**Reply by mkrueger** [reply]
y reverted some changes I did - messed things up. TIP should work again.

But I've got it running in a browser - to a degree:

![grafik](https://github.com/mkrueger/icy_term/assets/341098/2e566865-ceb8-4275-b681-4929ae20de8b)

**Reply by NuSkooler** [reply]
> y reverted some changes I did - messed things up. TIP should work again.

Did I push something bad?

> But I've got it running in a browser - to a degree:

Awesome progress!

**Reply by NuSkooler** [reply]
Update: Still having the drawing issue - tip of `no_tokio`, `main` icy_engine

**Reply by mkrueger** [reply]
y found another bug - a refactoring had a side effect did overlook a '?' so every parse error produced an issue.  I ran my manual tests now - looks good so far. 
I'll continue tomorrow. I need to finish the down & upload experience. That basically blocks that branch from beeing the new main. I think the communication is ok. 

I'll stabilize the drawing routines first - I added some opengl debug code that shows issues. ATM there are some bugs still left which  block the rendering engine from working it really sucks - it's maybe 5-10 LOC missing :). 

OpenGL is a nightmare. I don't want that at all - I just want shaders - but it's asked for too much :).

No one knows anything about these stupid error messages - I messed up something with frame buffers.

**Reply by NuSkooler** [reply]
> OpenGL is a nightmare. I don't want that at all - I just want shaders - but it's asked for too much :).

You're doing better than I!

Same issue still. I probably can't tonight, but should be able to look in the morning

**Comment by mkrueger** [comment]
Baud emulation is now working. 

Questionable is if we need an option for on the fly switching. Maybe just visible when the setting was turned on? 

And if we need upload baud emulation as well :). However it's always limited to what we get back.

**Comment by NuSkooler** [comment]
Yeah baud emulation should be working fine, though I keep having to 
re-fix icy_term for missing ps2 vs 0 :D

It needs to switch on the fly to work like boards expect, which 
basically emulates SyncTERM.

---
Bryan Ashby
www: https://l33t.codes/
PGP: https://keybase.io/nuskooler

On 2023-08-20 16:19, Mike Krüger wrote:

> Baud emulation is now working.
> 
> Questionable is if we need an option for on the fly switching. Maybe 
> just visible when the setting was turned on?
> 
> --
> Reply to this email directly, view it on GitHub [1], or unsubscribe 
> [2].
> You are receiving this because you authored the thread.Message ID: 
> ***@***.***>


Links:
------
[1] 
https://github.com/mkrueger/icy_term/discussions/19#discussioncomment-6773819
[2]
